pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java

```diff
@@ -0,0 +1,504 @@
+      _logger.info("Property 'hdfs.default.classpath.dir' not specified.");
+    }
+
+    long startTime = System.currentTimeMillis();
+    // Submit the job for execution.
+    job.waitForCompletion(true);
+    if (!job.isSuccessful()) {
+      throw new RuntimeException("Job failed : " + job);
+    }
+
+    _logger.info("Finished pre-computing job in {}ms", (System.currentTimeMillis() - startTime));
+  }
+
+  @Nullable
+  private TableConfig getTableConfig()
+      throws IOException {
+    try (ControllerRestApi controllerRestApi = getControllerRestApi()) {
+      return controllerRestApi != null ? controllerRestApi.getTableConfig() : null;
+    }
+  }
+
+  /**
+   * Can be overridden to provide custom controller Rest API.
+   */
+  @Nullable
+  private ControllerRestApi getControllerRestApi() {
+    return _pushLocations != null ? new DefaultControllerRestApi(_pushLocations, _rawTableName) : null;
+  }
+
+  private void setDirPermission(Path dirPath)
+      throws IOException {
+    if (_defaultPermissionsMask != null) {
+      FsPermission permission = FsPermission.getDirDefault().applyUMask(new FsPermission(_defaultPermissionsMask));
+      _logger.info("Setting permission: {} to directory: {}", permission, dirPath);
+      _fileSystem.setPermission(dirPath, permission);
+    }
+  }
+
+  private void fetchPartitioningConfig() {
+    // Fetch partition info from table config.
+    SegmentPartitionConfig segmentPartitionConfig = _tableConfig.getIndexingConfig().getSegmentPartitionConfig();
+    if (segmentPartitionConfig != null) {
+      Map<String, ColumnPartitionConfig> columnPartitionMap = segmentPartitionConfig.getColumnPartitionMap();
+      assert columnPartitionMap.size() <= 1;
+      if (columnPartitionMap.size() == 1) {
+        _partitionColumn = columnPartitionMap.keySet().iterator().next();
+        _numberOfPartitions = segmentPartitionConfig.getNumPartitions(_partitionColumn);
+        _partitionFunction = segmentPartitionConfig.getFunctionName(_partitionColumn);
+      }
+    } else {
+      _logger.info("Segment partition config is null for table: {}", _tableConfig.getTableName());
+    }
+
+    // Check whether configs overwritten by Hadoop properties.
+    if (_properties.getProperty(PARTITION_COLUMN) != null) {
+      _logger.info("Overriding partition column to: {}", _properties.getProperty(PARTITION_COLUMN));
+      _partitionColumn = _properties.getProperty(PARTITION_COLUMN);
+    }
+    if (_properties.getProperty(NUMBER_OF_PARTITIONS) != null) {
+      _logger.info("Overriding partition number to: {}", _properties.getProperty(NUMBER_OF_PARTITIONS));
+      _numberOfPartitions = Integer.parseInt(_properties.getProperty(NUMBER_OF_PARTITIONS));
+    }
+  }
+
+  private void fetchSortingConfig() {
+    // Fetch sorting info from table config.
+    IndexingConfig indexingConfig = _tableConfig.getIndexingConfig();
+    List<String> sortedColumns = indexingConfig.getSortedColumn();
+    assert sortedColumns.size() <= 1;
```

Comment:
you sholud throw an informative error for clients

Reply:
also, another problem is that clients will declare sorted columns today but not expect that we sort in their hadoop job still, so we may need to inform everyone

Reply:
If users don't want sort yet, they can still disable the preprocessing job. Plus, it's disabled by default.

Reply:
hmmm, i am thinking along the lines of when i want to resize by default, then i will enable preprocessing by default - do you think we should just turn sorting on for everyone? I think it will break, esp because some users think we can sort more than one column

Reply:
These 3 features won't affect each other; they can be enabled respectively. If you need resizing, just make `enable.resizing` to be true.
And we only allow one sorted column per table. Same as partition column.